analyzers/tests/SonarAnalyzer.TestFramework/Verification/IssueValidation/IssueLocation.cs

```diff
@@ -52,17 +56,25 @@ private IssueLocation(string message, Location location)
         FilePath = span.Path ?? string.Empty;   // Project-level issues do not have location
     }
 
+    public void UpdatePrimaryIssueIdFrom(IssueLocation expected)
+    {
+        if (IsPrimary)
+        {
+            issueId = expected.IssueId;  // Let actual secondary issues find issueId of their expected primary via secondaryIssue.Primary.IssueId
+        }
+    }
+
     public override int GetHashCode() =>
         Helpers.HashCode.Combine(FilePath.GetHashCode(), LineNumber, IsPrimary ? 0 : 1);
 
     public override bool Equals(object obj) =>
         obj is IssueLocation issue
-        && issue.FilePath == FilePath
+        && (issue.FilePath == string.Empty || FilePath == string.Empty || issue.FilePath == FilePath)
         && issue.LineNumber == LineNumber
         && issue.IsPrimary == IsPrimary
+        && (IsPrimary || issue.IssueId == IssueId)   // We ignore issueId for primary issues, as we need to match them only for secondary
         && (issue.RuleId is null || RuleId is null || issue.RuleId == RuleId)
         && (issue.Message is null || Message is null || issue.Message == Message)
-        && (issue.IssueId is null || IssueId is null || issue.IssueId == IssueId)   // For consistency, Roslyn issues will not have IDs
         && (issue.Start is null || Start is null || issue.Start == Start)
         && (issue.Length is null || Length is null || issue.Length == Length);
```

Comment:
I think we should add two helper methods here to make the big chain more readable:

```csharp
        private bool EqualOrNull(string first, string second) =>
            first is null || second is null || first == second;

        private bool EqualOrEmpty(string first, string second) =>
            first == string.Empty || second == string.Empty || first == second;
```